crates/bevy_animation/src/animator.rs

```diff
@@ -0,0 +1,1411 @@
+
+/// Defines the number of keyframes that fits inside a cache line;
+const KEYFRAMES_PER_CACHE: usize = CACHE_SIZE / std::mem::size_of::<u16>();
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct Tracks<T> {
+    outputs: SmallVec<[u16; 8]>,
+    /// Tuple of track and it's index
+    tracks: Vec<(usize, TrackBase<T>)>,
+    /// Similar to `tracks` but simultaneously has many outputs
+    ///
+    /// **NOTE** It has its own `outputs` stored inside
+    n: Vec<(usize, TrackNBase<T>)>,
+}
+
+impl<T> fmt::Debug for Tracks<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Tracks").field(&self.len()).finish()
+    }
+}
+
+impl<T> Tracks<T> {
+    /// Possible expensive function that calculates from scratch the total tracks duration;
+    ///
+    /// **NOTE** Caching the result value is desired
+    fn calculate_duration(&self) -> f32 {
+        self.tracks
+            .iter()
+            .map(|(_, track)| track.duration())
+            .chain(self.n.iter().map(|(_, track)| track.duration()))
+            .fold(0.0f32, |acc, d| acc.max(d))
+    }
+
+    /// Number of curves inside
+    pub fn len(&self) -> usize {
+        self.tracks.len() + self.n.iter().map(|(_, t)| t.len()).sum::<usize>()
+    }
+
+    #[inline(always)]
+    pub fn iter(
+        &self,
+    ) -> (
+        impl Iterator<Item = (u16, &(usize, TrackBase<T>))>,
+        &[(usize, TrackNBase<T>)],
+    ) {
+        (
+            self.outputs.iter().copied().zip(self.tracks.iter()),
+            &self.n,
+        )
+    }
+}
+
+pub struct TrackMeta(&'static str, TypeId);
+
+impl TrackMeta {
+    pub fn of<T: 'static>() -> Self {
+        Self(type_name::<T>(), TypeId::of::<T>())
+    }
+
+    pub const fn type_name(&self) -> &str {
+        self.0
+    }
+
+    pub const fn type_id(&self) -> TypeId {
+        self.1
+    }
+}
+
+impl fmt::Debug for TrackMeta {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("TrackMeta").field(&self.0).finish()
+    }
+}
+
+#[derive(Debug)]
+pub struct TracksUntyped {
+    /// Cached calculated curves duration
+    duration: f32,
+    meta: TrackMeta,
+    untyped: Box<dyn Any + Send + Sync + 'static>,
+}
+
+impl TracksUntyped {
+    fn new<T: Send + Sync + 'static>(tracks: Tracks<T>) -> Self {
+        TracksUntyped {
+            duration: tracks.calculate_duration(),
+            meta: TrackMeta::of::<T>(),
+            untyped: Box::new(tracks),
+        }
+    }
+
+    pub const fn meta(&self) -> &TrackMeta {
+        &self.meta
+    }
+
+    pub const fn duration(&self) -> f32 {
+        self.duration
+    }
+
+    #[inline(always)]
+    pub fn downcast_ref<T: 'static>(&self) -> Option<&Tracks<T>> {
+        self.untyped.downcast_ref()
+    }
+
+    #[inline(always)]
+    pub fn downcast_mut<T: 'static>(&mut self) -> Option<&mut Tracks<T>> {
+        self.untyped.downcast_mut()
+    }
+}
+
+// Pros:
```

Comment:
I feel like this should be part of the PR draft. It has a lot of useful information and estimations cart should see upfront.

Reply:
I wanted that the animation system to be the sole responsible of finding and assigning any animated entities, so I opted for using a Hierarchy instead of a entity list;

Reply:
I mean more moving the Pros and Cons into the PR draft.md, so people can pick up quickly on what you took into consideration and what compromises you have chosen.